### PR TITLE
feat(static): Improvements to Cape PDF static extraction:

### PR DESCRIPTION
**Pull request by yoniabrahamy**
* Add PDF miner for better URI extraction
* Add indicator for PDF containing Javascript

### CODE REVIEW

File: modules/processing/static.py [hunk 1]
```diff
@@ -1109,6 +1145,40 @@ def _set_base_uri(self):
             log.error(e, exc_info=True)
             pass
 
+    @staticmethod
+    def _search_for_url(obj: Union[pdftypes.PDFStream, dict, list]):
```
Comment (yoniabrahamy) [inline modules/processing/static.py]: This is where the changes really start, all the rest is mostly formatting

Comment (davidt99) [inline modules/processing/static.py]: what does it returns?

File: modules/processing/static.py [hunk 2]
```diff
@@ -1109,6 +1145,40 @@ def _set_base_uri(self):
...
+        if obj is None:
+            return
+
+        if isinstance(obj, pdftypes.PDFStream):
+            yield from PDF._search_for_url(obj.attrs)
+        elif isinstance(obj, list):
+            for v in obj:
```
Comment (davidt99) [inline modules/processing/static.py:1158]: can we have better variables names?

File: modules/processing/static.py [hunk 3]
```diff
@@ -1109,6 +1145,40 @@ def _set_base_uri(self):
...
+                yield from PDF._search_for_url(v)
+        elif isinstance(obj, dict):
+            for k, v in obj.items():
+                if k == 'URI':
+                    yield v
+                yield from PDF._search_for_url(v)
```
Comment (davidt99) [inline modules/processing/static.py]: Wouldn't we yield the same value twice in case of a URI?

File: modules/processing/static.py [hunk 4]
```diff
@@ -1109,6 +1145,40 @@ def _set_base_uri(self):
...
+                parser = pdfparser.PDFParser(f)
+                doc = pdfdocument.PDFDocument(parser)
+
+                for xref in doc.xrefs:
+                    for object_id in xref.get_objids():
+                        obj = doc.getobj(object_id)
+                        urls.update(PDF._search_for_url(obj))
+        except Exception as ex:
+            log.error(ex, exc_info=True)
+            log.warning('Failed opening file for pdf mining')
```
Comment (davidt99) [inline modules/processing/static.py]: why twice the logs?

File: modules/processing/static.py [hunk 5]
```diff
@@ -1109,6 +1145,40 @@ def _set_base_uri(self):
...
+        elif isinstance(obj, dict):
+            for k, v in obj.items():
+                if k == 'URI':
+                    yield v
+                yield from PDF._search_for_url(v)
+
+    @staticmethod
+    def _mine_for_urls(file_path: str):
+        urls = set()
+        try:
```
Comment (davidt99) [inline modules/processing/static.py:1171]: the try and except can be inside the for as well?

File: modules/processing/static.py [hunk 6]
```diff
@@ -1244,11 +1314,19 @@ def _parse(self, filepath):
...
+            pdfresult["Info"]["JavaScript"] = "True"
+
         if len(urlset):
             pdfresult["JS_URLs"] = list(urlset)
         if len(annoturiset):
             pdfresult["Annot_URLs"] = list(annoturiset)
 
+        if HAVE_PDFMINER:
+            mined_urls = set((self.base_uri + url.decode() for url in self._mine_for_urls(self.file_path)))
+            if len(mined_urls):
```
Comment (davidt99) [inline modules/processing/static.py]: why do we need the len?
```suggestion
            if mined_urls:
```

File: modules/processing/static.py [hunk 7]
```diff
@@ -1244,11 +1314,19 @@ def _parse(self, filepath):
...
+        if pdfresult["Keywords"].get("/JS", 0) > 0 or pdfresult["Keywords"].get("/JavaScript", 0) > 0:
+            pdfresult["Info"]["JavaScript"] = "True"
+
         if len(urlset):
             pdfresult["JS_URLs"] = list(urlset)
         if len(annoturiset):
             pdfresult["Annot_URLs"] = list(annoturiset)
 
+        if HAVE_PDFMINER:
+            mined_urls = set((self.base_uri + url.decode() for url in self._mine_for_urls(self.file_path)))
```
Comment (davidt99) [inline modules/processing/static.py]:
```suggestion
            mined_urls = set(self.base_uri + url.decode() for url in self._mine_for_urls(self.file_path))
```

**Review by davidt99 — APPROVED** [review]

**Pull request merged**